GraphicsContext::drawLinesForText might read out of bounds

kkinnunen-apple opened the pull request:
#### 15cc8962394fa5f0ebb029ff188f47f07d841360
<pre>
GraphicsContext::drawLinesForText might read out of bounds
<a href="https://bugs.webkit.org/show_bug.cgi?id=286570">https://bugs.webkit.org/show_bug.cgi?id=286570</a>
<a href="https://rdar.apple.com/143270489">rdar://143270489</a>

Reviewed by Antti Koivisto.

A compromised WCP could send width array with uneven amount of elements
to GraphicsContext::drawLinesForText(). This would cause out-of-bounds
read, as drawLinesForText expects [begin, end] pairs of line segments.

The root cause is that the function used incorrect type DashArray
to pass the line segments. DashArray is a type to pass doubles
that represent lengths relative to previous elements. When used
correctly DashArray [5,6,7] would encode segment list [{0, 5}, {11, 18}].
Note, that the implementation did not use DashArray this way.

Fix by introducing FloatSegment and pass Vector&lt;FloatSegment&gt; instead
of DashArray.

This fixes following issues:
- Avoid passing doubles. The source algorithm works
  with floats, so sending doubles is redundant.
- Avoid conversion from float pair array to float array in the source
  algorithm. The source algorithm already works with float pairs.

Since the change of types touches all of the lines in the source
algorithm, do also following:
- Simplify source algorithm
- Avoid copies of the vectors in the source algorithm

Source algorithm would do following full passes:
 1. Dilate intersection line segments
 2. Sort interections along segment begin
 3. Merge overlapping intersection segments
 4. Compute difference between a solid line and intersections

 Instead, do following full passes:
 1. Sort intersections along segment begin
 2. Compute difference between a solid line and intersections, with
   intersections dilated.

The difference can be computed with overlapping intersections, overlaps
do not affect the result.

For GraphicsContext functions:
- Pass std::span&lt;const FloatSegment&gt; instead of Vector&lt;FloatSegment&gt;,
  as the callee does not need to use Vector functionality anywhere.
- Pass span as value, because it&apos;s small and this prevents pointer
  chasing.
- Fix few .reserveInitialCapacity() size calculations, previously
  tried to always reserve 0.
- Share the rect conversion implementation with Skia port. Skia had
  buggy variant of of the rect conversion algorithm cut-pasted from
  CG variant before CG variant was improved.

Remove all default arguments from GraphicsContext::drawLinesForText().
It is poor style to have default arguments for virtual functions, as
then all overrrides need to replicate them in order to guarantee
consistent invocation through derived type pointer. The callers did
not use the default arguments.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/BifurcatedGraphicsContext.cpp:
(WebCore::BifurcatedGraphicsContext::drawLinesForText):
* Source/WebCore/platform/graphics/BifurcatedGraphicsContext.h:
* Source/WebCore/platform/graphics/DashArray.h:
* Source/WebCore/platform/graphics/FloatSegment.cpp: Copied from Source/WebCore/platform/graphics/DashArray.h.
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/graphics/FloatSegment.h: Added.
(WebCore::FloatSegment::length const):
(WebCore::FloatSegment::dilate):
(WebCore::differenceWithDilation):
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::lineSpansForIntersectionsWithRect const):
(WebCore::FontCascade::dashesForIntersectionsWithRect const): Deleted.
* Source/WebCore/platform/graphics/FontCascade.h:
* Source/WebCore/platform/graphics/GraphicsContext.cpp:
(WebCore::GraphicsContext::drawLineForText):
(WebCore::GraphicsContext::computeRectsAndStrokeColorForLinesForText):
* Source/WebCore/platform/graphics/GraphicsContext.h:
* Source/WebCore/platform/graphics/NullGraphicsContext.h:
* Source/WebCore/platform/graphics/cairo/CairoOperationRecorder.cpp:
(WebCore::Cairo::OperationRecorder::drawLinesForText):
* Source/WebCore/platform/graphics/cairo/CairoOperationRecorder.h:
* Source/WebCore/platform/graphics/cairo/CairoOperations.cpp:
(WebCore::Cairo::drawLinesForText):
* Source/WebCore/platform/graphics/cairo/CairoOperations.h:
* Source/WebCore/platform/graphics/cairo/GraphicsContextCairo.cpp:
(WebCore::GraphicsContextCairo::drawLinesForText):
* Source/WebCore/platform/graphics/cairo/GraphicsContextCairo.h:
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::GraphicsContextCG::drawLinesForText):
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp:
(WebCore::DisplayList::DrawLinesForText::DrawLinesForText):
(WebCore::DisplayList::DrawLinesForText::apply const):
(WebCore::DisplayList::DrawLinesForText::dump const):
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.h:
(WebCore::DisplayList::DrawLinesForText::lineSegments const):
(WebCore::DisplayList::DrawLinesForText::widths const): Deleted.
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.cpp:
(WebCore::DisplayList::RecorderImpl::drawLinesForText):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.h:
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
(WebCore::GraphicsContextSkia::drawLinesForText):
* Source/WebCore/rendering/TextDecorationPainter.cpp:
(WebCore::TextDecorationPainter::paintBackgroundDecorations):
(WebCore::compareTuples): Deleted.
(WebCore::translateIntersectionPointsToSkipInkBoundaries): Deleted.
* Source/WebKit/Shared/DisplayListArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::drawLinesForText):
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.h:
* Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
* Tools/TestWebKitAPI/CMakeLists.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/FloatSegmentTest.cpp: Added.
(TestWebKitAPI::TEST(FloatSegment, DilateWithDifference)):

Canonical link: <a href="https://commits.webkit.org/289737@main">https://commits.webkit.org/289737@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d82cf01e7692c52039f99b468e9752f63906900a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87825 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7341 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42213 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92694 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38575 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89876 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7722 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15514 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67817 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25559 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90827 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5907 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79467 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48185 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5694 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33875 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37682 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76092 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34754 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94583 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14993 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/11033 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76661 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15248 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75323 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75897 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18663 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20271 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18707 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7997 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/15009 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/20312 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14753 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18197 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16535 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/platform/graphics/FloatSegment.h:
```diff
@@ -0,0 +1,81 @@
+// return filter(diff, [](auto&& s) { return s.length() > dilationAmount });
+// <=>
+// return differenceWithDilation({ 0, totalWidth }, intersections);
+inline Vector<FloatSegment> differenceWithDilation(FloatSegment a, Vector<FloatSegment>&& bs, float dilationAmount)
```
I don't know if "dilation" is a term of art but there could be more explanation what this actually achieves, for example with some ascii art with lines and segments.

The pseudo-code is confusing, it is not clear `-` is a set operator for example.

anttijk commented on Source/WebCore/platform/graphics/FloatSegment.h:
```diff
@@ -0,0 +1,81 @@
+        return a.begin < b.begin;
+    });
+
+    auto diff = bs.begin();
```
Not clear why this is called 'diff' since it is not a difference of any sort.

anttijk commented on Source/WebCore/platform/graphics/FloatSegment.h:
```diff
@@ -0,0 +1,81 @@
+    bs.shrink(diff - bs.begin());
+    if (a.length() > dilationAmount)
+        bs.append(a);
+    return WTFMove(bs);
```
This algorithm would probably be clearer if it just created a new vector instead of mutating the existing one in-place.

kkinnunen-apple committed:
GraphicsContext::drawLinesForText might read out of bounds
https://bugs.webkit.org/show_bug.cgi?id=286570
rdar://143270489

Reviewed by Antti Koivisto.

A compromised WCP could send width array with uneven amount of elements
to GraphicsContext::drawLinesForText(). This would cause out-of-bounds
read, as drawLinesForText expects [begin, end] pairs of line segments.

The root cause is that the function used incorrect type DashArray
to pass the line segments. DashArray is a type to pass doubles
that represent lengths relative to previous elements. When used
correctly DashArray [5,6,7] would encode segment list [{0, 5}, {11, 18}].
Note, that the implementation did not use DashArray this way.

Fix by introducing FloatSegment and pass Vector<FloatSegment> instead
of DashArray.

This fixes following issues:
- Avoid passing doubles. The source algorithm works
  with floats, so sending doubles is redundant.
- Avoid conversion from float pair array to float array in the source
  algorithm. The source algorithm already works with float pairs.

Since the change of types touches all of the lines in the source
algorithm, do also following:
- Simplify source algorithm
- Avoid copies of the vectors in the source algorithm

Source algorithm would do following full passes:
 1. Dilate intersection line segments
 2. Sort interections along segment begin
 3. Merge overlapping intersection segments
 4. Compute difference between a solid line and intersections

 Instead, do following full passes:
 1. Sort intersections along segment begin
 2. Compute difference between a solid line and intersections, with
   intersections dilated.

The difference can be computed with overlapping intersections, overlaps
do not affect the result.

For GraphicsContext functions:
- Pass std::span<const FloatSegment> instead of Vector<FloatSegment>,
  as the callee does not need to use Vector functionality anywhere.
- Pass span as value, because it's small and this prevents pointer
  chasing.
- Fix few .reserveInitialCapacity() size calculations, previously
  tried to always reserve 0.
- Share the rect conversion implementation with Skia port. Skia had
  buggy variant of of the rect conversion algorithm cut-pasted from
  CG variant before CG variant was improved.

Remove all default arguments from GraphicsContext::drawLinesForText().
It is poor style to have default arguments for virtual functions, as
then all overrrides need to replicate them in order to guarantee
consistent invocation through derived type pointer. The callers did
not use the default arguments.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/BifurcatedGraphicsContext.cpp:
(WebCore::BifurcatedGraphicsContext::drawLinesForText):
* Source/WebCore/platform/graphics/BifurcatedGraphicsContext.h:
* Source/WebCore/platform/graphics/DashArray.h:
* Source/WebCore/platform/graphics/FloatSegment.cpp: Copied from Source/WebCore/platform/graphics/DashArray.h.
(WebCore::operator<<):
* Source/WebCore/platform/graphics/FloatSegment.h: Added.
(WebCore::FloatSegment::length const):
(WebCore::FloatSegment::dilate):
(WebCore::differenceWithDilation):
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::lineSpansForIntersectionsWithRect const):
(WebCore::FontCascade::dashesForIntersectionsWithRect const): Deleted.
* Source/WebCore/platform/graphics/FontCascade.h:
* Source/WebCore/platform/graphics/GraphicsContext.cpp:
(WebCore::GraphicsContext::drawLineForText):
(WebCore::GraphicsContext::computeRectsAndStrokeColorForLinesForText):
* Source/WebCore/platform/graphics/GraphicsContext.h:
* Source/WebCore/platform/graphics/NullGraphicsContext.h:
* Source/WebCore/platform/graphics/cairo/CairoOperationRecorder.cpp:
(WebCore::Cairo::OperationRecorder::drawLinesForText):
* Source/WebCore/platform/graphics/cairo/CairoOperationRecorder.h:
* Source/WebCore/platform/graphics/cairo/CairoOperations.cpp:
(WebCore::Cairo::drawLinesForText):
* Source/WebCore/platform/graphics/cairo/CairoOperations.h:
* Source/WebCore/platform/graphics/cairo/GraphicsContextCairo.cpp:
(WebCore::GraphicsContextCairo::drawLinesForText):
* Source/WebCore/platform/graphics/cairo/GraphicsContextCairo.h:
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::GraphicsContextCG::drawLinesForText):
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp:
(WebCore::DisplayList::DrawLinesForText::DrawLinesForText):
(WebCore::DisplayList::DrawLinesForText::apply const):
(WebCore::DisplayList::DrawLinesForText::dump const):
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.h:
(WebCore::DisplayList::DrawLinesForText::lineSegments const):
(WebCore::DisplayList::DrawLinesForText::widths const): Deleted.
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.cpp:
(WebCore::DisplayList::RecorderImpl::drawLinesForText):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.h:
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
(WebCore::GraphicsContextSkia::drawLinesForText):
* Source/WebCore/rendering/TextDecorationPainter.cpp:
(WebCore::TextDecorationPainter::paintBackgroundDecorations):
(WebCore::compareTuples): Deleted.
(WebCore::translateIntersectionPointsToSkipInkBoundaries): Deleted.
* Source/WebKit/Shared/DisplayListArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::drawLinesForText):
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.h:
* Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
* Tools/TestWebKitAPI/CMakeLists.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/FloatSegmentTest.cpp: Added.
(TestWebKitAPI::TEST(FloatSegment, DilateWithDifference)):

Canonical link: https://commits.webkit.org/289737@main
